### ISSUE TITLE
Restrictive time window for circuit breaker to record failures

### ISSUE THREAD

**Issue by whiskeysierra** [jhalterman/failsafe#208]
Right now it's possible to configure failure thresholds in terms of consecutive failures:

- `withFailureThreshold(4, 5)`, four failures out of five consecutive executions
- `withFailureThreshold(5, 10)`, five failures out of ten consecutive executions
- etc.

There is no notion of time right now. What I'm hoping for would be the ability to specify something like:
- *10 failures within 1 second*
- *50 failures out of 200 executions within a minute*

My idea behind this is to effectively disable the circuit breaker in low-traffic scenarios but if traffic suddenly increases it should start to kick in. For requests with low traffic (less than ~10 per second) we identified circuit breakers as actually making matters worse, since they tend to stay open for longer, once they open up. And ultimately circuit breaker are means to protect overloading an application which doesn't really happen with few requests.

**Comment by Tembrel** [comment]
I feel as though it would almost be possible to compose a useful subset of the proposed behavior out of existing policies.

But I agree with the motivating thought: "Last 4 out of 5 attempts failed" can mean very different things if those 5 were spread out over seconds rather than milliseconds, at least in some contexts.

**Comment by whiskeysierra** [comment]
Not sure if this is what I'm looking for but resilience4j has something that sounds promising: [Time-based sliding window](https://resilience4j.readme.io/docs/circuitbreaker#section-time-based-sliding-window)

**Comment by jhalterman** [comment]
Consolidating the discussion with #214, I'd like to nail down the various aspects of this feature. Please share your thoughts on the items below @whiskeysierra @armelbd @Tembrel.

### Use cases

The use cases described above and in #214 sound pretty reasonable, and seem very easy to understand. They include:

1. Certain number of failures over a time period (ex, 10 executions failed in the last minute)
2. Certain number of executions failed over some time period (ex, 5 out of last 10 executions failed in the last minute)
3. Certain ratio of executions failed over some time period (ex: 20% of executions failed in the last minute)
  a. This feature is potentially problematic since the threshold could easily be exceeded early on. For example, a 20% ratio would be exceeded if any of the first 5 calls fail. To make this useful we should either require a minimum number of calls that must be executed, or, compute a reasonable minimum based on the given ratio (ex: 1 / ratio). Thoughts?

Let me know which if the above use cases make sense and if I'm missing any.

### Performance considerations

In order to avoid storing the time of every execution and doing a lot of list traversal to regularly purge old executions, it makes sense to aggregate execution results into buckets that users can help size. It's unfortunate that users should have to even worry about this, but it does let them choose their own tradeoff between execution result resolution and the time/space to maintain them.

Ultimately there are 3 variables involved in sizing: total size of the window (time), number of buckets, and size of each bucket (time). In practice, only 1-2 of them need to be configured. Some options:

- a fixed bucket size of 1 second and number of buckets
  - Ex: 30 buckets = 30 second window
- bucket size (time) and number of buckets
  - Ex: 6 buckets, 10 second bucket size = 60 second window
- window size (time) and number of buckets
  - Ex: 30 second window, 100 buckets = .3 second bucket size (resolution)

Which approach we choose depends on where we want users to focus more, on the total time period or on the bucket size. The use cases above focus more on the total time period, which makes sense to me. This means bucket sizes may end up being arbitrary (ex: .3 seconds), but that doesn't really matter. Thoughts on this?

### API considerations

As far as API, I think there are two general approaches. Add some new `withFailureThreshold` methods. Ex:

- `withFailureThreshold(int failures, Duration timePeriod)`
- `withFailureThreshold(int failures, int executions, Duration timePeriod)`
- `withFailureThreshold(double failureRatio, Duration timePeriod)`
- `withFailureThreshold(double failureRatio, int minimumExecutions, Duration timePeriod)`
-`withFailureThreshold(int failures, Duration timePeriod, int buckets)`
...etc

There could end up being quite a lot of these to cover all permutations. An alternative might be to combine the existing `withFailureThreshold` methods with a new method that captures just the time-based configuration info. Ex:

- `withPeriod(Duration timePeriod, int buckets)`

Thoughts?

## Implementation notes

CircuitBreaker currently uses `CircularBitSet` internally to track whether the last N executes succeeded or failed. For a sliding window, this data structure won't quite work as is. Instead, we'll need to track the number of failures and executions for each of the last N seconds/minutes.

**Comment by whiskeysierra** [comment]
> it makes sense to aggregate execution results into buckets that users can help size

This would be similiar to resilience4j which uses sliding windows (either count or time based).
https://resilience4j.readme.io/docs/circuitbreaker

Maybe it would be best to just accept a `SlidingWindow` (interface) as a parameter and let users pass one of two implementations that Failsafe supports? That would help to keep the API small, also:

> There could end up being quite a lot of these to cover all permutations.

Maybe it's worth to consider a similar approach as Google Flogger?
https://google.github.io/flogger/anatomy

**Comment by jhalterman** [comment]
> This would be similiar to resilience4j which uses sliding windows 

I did have a look. What I don't like about the approach there is it focuses on the underlying data structure rather than the thing a user actually wants to accomplish, ex: threshold on 5 failed executions in 1 minute. The internals should be secondary to what the user wants to accomplish.

> Maybe it's worth to consider a similar approach as Google Flogger?

You mean with the fluent DSL, or something else?

**Comment by whiskeysierra** [comment]
> Of course you still may have a bucketed sliding window either way, but the internals should be secondary to focusing on what the user wants to accomplish.

Agreed. From an implementation perspective sliding windows is probably what we need. How to expose it as an API is a different concern.

> You mean with the fluent DSL, or something else?

Exactly. Maybe the combinatorial explosion can be mitigated by having a smart, fluent API.

**Comment by jhalterman** [comment]
Spending some more time thinking about this today, I think we can get away with just two sets of methods to configure this. The first define the threshold, min executions before the breaker can be opened, and the time period:

```java
withFailureThreshold(failures, timePeriod) // minExecutions = failures
withFailureThreshold(failures, minExecutions, timePeriod)
withFailureThreshold(failureRatio, minExecutions, timePeriod)
```

Then an optional method can specify a value for the sliding window's resolution, which basically sub-divides the internal sampling into some number of buckets:

```java
withResolution(10) // divide timePeriod into 10 buckets
```

The larger the resolution, the smoother and more accurate the transition between buckets as the window moves.

We will need to choose a default value for the resolution. Some other defaults I've seen in the wild:
- 6 in Hystrix (which corresponds to a default timePeriod of 60, each bucket represents 10 seconds)
- 10 in Polly (buckets can represent varying time periods)
- 100 in Resilience4j (which equals the default timePeriod of 100, buckets always represent 1 second)

The consideration here is that a smaller number means we discard more data at a time, meaning the overall measurement is less accurate. For example, Hystrix's default of 6 buckets over 60 seconds means each bucket represents 10 seconds. So as the window moves we'd be discarding ~16% of the data at a time. With 10 buckets representing 6 seconds each, we'd be discarding 10% of the data at a time.

In terms of choosing a really large number, performance will be impacted. A greater number means more samples to store, and also more iterations to perform as each time an execution is performed we may need to iterate through the list of buckets to remove old ones. 

That all said, I'm wondering if we should choose a fixed default, something in the 10-20 range, or let the default be based on the timePeriod? The latter would allow us to provide more buckets for larger windows, but we'd still want to use a function that doesn't produce too many buckets - something logarithmic. 

Thoughts on all this?

**Comment by armelbd** [comment]
> This feature is potentially problematic since the threshold could easily be exceeded early on. For example, a 20% ratio would be exceeded if any of the first 5 calls fail.

Very good point, we need something to address it, this makes a lot of sense:


```Java
withFailureThreshold(failures, timePeriod) // minExecutions = failures
withFailureThreshold(failures, minExecutions, timePeriod)
```

I'm less sure about:
```Java
withFailureThreshold(failureRatio, minExecutions, timePeriod)
```

As I was mentioning in #214, number of executions is not always something we control, if a failure ratio is provided a more natural way would be:

```Java
withFailureThreshold(failureRatio, minPeriod, timePeriod)
withFailureThreshold(failureRatio, timePeriod) // minPeriod = timePeriod
```

Regarding the number of buckets, for users I believe it makes more sense to think about it has a time resolution (i.e how often the window move). Maybe by setting it as a number of buckets it makes the performance cost a bit more transparent but it requires a sense of how it's implemented. I would be in favor of setting it like that:

```Java
withResolution(timePeriod)
```
If nothing is provided by the user, than I guess it makes sense to base it on the size of the window and have a fix number of buckets, 10 seems reasonable.

Anyway, it's awesome to see this getting interest!

**Comment by jhalterman** [comment]
> withFailureThreshold(failureRatio, minPeriod, timePeriod)

What would `minPeriod` represent here?

> withResolution(timePeriod)

I was considering this as well... we could certainly support either way. Expressing things as a number of buckets is more typical, but in that case the width of the bucket is derived from the windowSize / numberOfBuckets.

**Comment by reisenberger** [comment]
>10 in Polly (buckets can represent varying time periods)

As you noted, in Polly, we chose 10 buckets as a fixed configuration.  FWIW, in the four years since the Polly version has been released, we haven't had any user request to make that configurable (greater or lesser number of buckets/resolution).

**Comment by jhalterman** [comment]
> we haven't had any user request to make that configurable (greater or lesser number of buckets/resolution).

That's really great to know, thanks for chiming in. I would expect the lack of smoothness with longer windows could be an issue. Do you think perhaps it is and users just aren't noticing?

**Comment by jhalterman** [comment]
@armelbd I wanted to follow up on this comment of yours above:

> As I was mentioning in #214, number of executions is not always something we control, if a failure ratio is provided a more natural way would be:
withFailureThreshold(failureRatio, minPeriod, timePeriod)
withFailureThreshold(failureRatio, timePeriod) // minPeriod = timePeriod


Doesn't that run the risk of the first failure exceeding the ratio?

**Comment by armelbd** [comment]
@jhalterman in:
`withFailureThreshold(failureRatio, minPeriod, timePeriod)`
`minPeriod` serves the same purpose as `minExecutions` in your example. Until this time has elapsed we won't enforce the failure ratio. If not provided we would default minPeriod to timePeriod. Does that make sense?

**Comment by jhalterman** [comment]
@armelbd It does make sense, it just still might allow for a few executions, even 1, to open the breaker (assuming `minPeriod` has been exceeded), regardless of what the `failureRatio` is.

**Comment by armelbd** [comment]
oh ok, got it. I have always use Failsafe in a context where there is a constant flow of executions, but I see that it's a valid concern in other scenarios. I still would want an `minPeriod` though, it would be useful in cases where an higher failure rate is expected at startup ("warming up" period)

**Comment by jhalterman** [comment]
Pushed another revision of this feature: cea14c4a5995c722b3a63aed3d408e929dc01491

I need to add some more tests but am nearing completion.

**Comment by jhalterman** [comment]
I just merged this to master: d2da5a5b9d54d9fb348f42407178c72d1529adf4. The docs for the website have been pushed here fc5353251313cecef9675088292aed7d9af19bb7.

If anyone has any comments on anything they think needs to be changed, feel free to share them now before I cut a release :)

There are a few bits I left out for now that can easily be added such as a configurable number of buckets and also `withFailureThreshold(thresh, minPeriod, period)`. Regarding the latter - @armelbd I'm curious how necessary that is for you? Would the `minPeriod` be less than the `failureThresholdingPeriod`?

**Comment by armelbd** [comment]
Thanks for this, really cool feature!

@jhalterman It's a nice to have, having `failureExecutionThreshold` is helpful but not 100% ideal. I don't think there is necessarily a relationship between `minPeriod` and `failureThresholdingPeriod`. I can imagine cases where I would use a `minPeriod` smaller than `failureThresholdingPeriod` and other cases where I would set it to a longer period. Looking at the implementation, I guess you are asking this because it would be easy to based it on the number of buckets? Making `minPeriod` a multiple of `failureThresholdingPeriod` would be good enough to me!

**Comment by jhalterman** [comment]
> Looking at the implementation, I guess you are asking this because it would be easy to based it on the number of buckets?

I'm just trying to understand this potential feature and when to use different values. I'm not sure if it makes sense if `minPeriod` were longer than `failureThresholdingPeriod`. That would mean that while we waited for `minPeriod` to elapse, we're only able to threshold on data from the `failureThresholdingPeriod` (because older data is discarded).

**Comment by jhalterman** [comment]
Time bases breakers have been released in 2.4. Check the docs:

- https://jodah.net/failsafe/circuit-breaker/#opening
- https://jodah.net/failsafe/circuit-breaker/#time-based-resolution

I'd like to close this issue and open a separate one to consider `withFailureThreshold(thresh, minPeriod, period)` specifically. We can consider that there in #250.
